02-express-tutorial/app.js

```diff
@@ -1 +1,66 @@
-console.log('Express Tutorial')
+const express = require ('express');
+const app = express ();
+const { products } = require('./data');
+
+// Setup static
+app.use(express.static ('./public' ));
+
+// API that return JSON
+app.get('/api/v1/test', (req, res) => {
+    res.json({message: 'It worked!'});
+})
+
+// API that returns products data
+app.get('/api/v1/products', (req, res) => {
+    res.json(products);
+});
+
+// API that returns products by ID
+app.get('/api/v1/products/:productID', (req, res) => {
+    const idToFind = parseInt(req.params.productID);
+    const singleProduct = products.find(
+        (product) => product.id === idToFind
+        );
+    if (!singleProduct) {
+        return res.status(404).json({ message: 'That product was not found.' })
+    }
+    return res.json(singleProduct);
+});
+
+// Query
+app.get('/api/v1/query', (req, res) => {
+    const { search, limit, priceLimit } = req.query;
+    let sortedProducts = [...products]
+
+    // Filter products
+    if (search) {
```

Comment:
Requires change:
If you notice, you're using 2 filters, one for `search` and one for `priceLimit`. This is not very efficient because you have to iterate over the entire products list twice. Instead, I'd recommend using [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) which will check all 3 conditions in ONE iteration.
```js
app.get('/api/v1/query', (req, res) => {
    const { search, limit = 0, priceLimit = 0 } = req.query;

    const priceLimit = parseInt(limit, 10);

    // use Array.reduce to build a list of filtered products
    const filteredProducts = products.reduce((acc, product) => {
        // if the product price is greater than priceLimit OR
        // if there's a search and the product name doesn't include the search term OR
        // if there’s a limit and the accumulator (acc === list of filtered products) hit the limit
        // then return the accumulator (acc === list of filtered products)
        if (
            product.price > parseFloat(priceLimit) || 
            (search && !product.name.includes(search)) ||
            (maxLimit && acc.length === maxLimit)
        ) {
            return acc;
        }
       
        // add product to accumulator list
        acc.push(product);

        // return the accumulator to check the next product
        return acc;
    }, []);

    res.status(200).json(filteredProducts);
});
```